Add Linux 5.1+ compat fix for deprecated CRYPTO_TFM_REQ_WEAK_KEY

ColinIanKing opened the pull request:
With Linux 5.1 onwards, CRYPTO_TFM_REQ_WEAK_KEY has been deprecated and
renamed to CRYPTO_TFM_REQ_FORBID_WEAK_KEYS.  Add a compat check to cater
for this change as introduced into Linux with commit 231baecdef7a9065799
("crypto: clarify name of WEAK_KEY request flag")

Signed-off-by: Colin Ian King <colin.king@canonical.com>

lumag commented on gost28147_modes.c:
```diff
@@ -282,7 +282,7 @@ static int gost28147_cfb_decrypt_inplace(struct skcipher_walk *walk,
 	} while ((nbytes -= bsize) >= bsize);
 
-	memcpy(walk->iv, iv, bsize);
+	memmove(walk->iv, iv, bsize);
```
Hmm. Mainline linux CFB code doesn't contain this memcpy. I should probably drop it alltogether. Would you mind backporting  6c2e322b3621dc8be72e5c86d4fdb587434ba625 instead?

lumag commented:
As for the 5.1 I've included a bit different fix. Could you please check whether it works for you?

ColinIanKing commented:
> As for the 5.1 I've included a bit different fix. Could you please check whether it works for you?

Apologies, but I can't see the fix you are referring to. Any hint which commit that is?

Commit:
crypto: cfb - remove bogus memcpy() with src == dest

The memcpy() in crypto_cfb_decrypt_inplace() uses walk->iv as both the
source and destination, which has undefined behavior.  It is unneeded
because walk->iv is already used to hold the previous ciphertext block;
thus, walk->iv is already updated to its final value.  So, remove it.

Also, note that in-place decryption is the only case where the previous
ciphertext block is not directly available.  Therefore, as a related
cleanup I also updated crypto_cfb_encrypt_segment() to directly use the
previous ciphertext block rather than save it into walk->iv.  This makes
it consistent with in-place encryption and out-of-place decryption; now
only in-place decryption is different, because it has to be.

Based on upstream linux commit 6c2e322b3621 ("crypto: cfb - remove bogus
memcpy() with src == dest")

Fixes: f816e4949591 ("gost28147: add CFB support with key meshing")
Signed-off-by: Colin Ian King <colin.king@canonical.com>

ColinIanKing commented:
I've force pushed the backport of commit  crypto: cfb - remove bogus memcpy() with src == dest to my repo and requesting a merge of this.

lumag commented on gost28147_modes.c:
```diff
@@ -158,11 +158,14 @@ static int gost28147_cfb_encrypt_segment(struct skcipher_walk *walk,
 		gost28147_cfb_encrypt_one(ctx, iv, dst);
 		crypto_xor(dst, src, bsize);
 		memcpy(iv, dst, bsize);
```
This line (`memcpy`) results in tests failure. Please remove it, as done in the original commit.

lumag commented:
> > As for the 5.1 I've included a bit different fix. Could you please check whether it works for you?
> 
> Apologies, but I can't see the fix you are referring to. Any hint which commit that is?

My fault, I've pushed it to the tag 0.3.1, but failed to push to master branch. Should be fixed now.

lumag commented:
@ColinIanKing I'll tag 0.3.2 after merging this pull request and upload it to Debian.

lumag commented:
Merged manually. Closing. Thanks @ColinIanKing
